Exposing the data definition epsilon variable plus fix example

BitsAndDroids opened the pull request:
## Example code fix
Fixed the example code to work with the latest Rust version regarding packed structs. The old example code couldn't compile anymore.

## Epsilon
When the epsilon of all variables in a data definition is set to 0.0, the sim will return an abundance of data when data is flagged as on change.
For example, every mm change of altitude is seen as a change in the variable, while in reality, you most likely only need changes per foot. Exposing the epsilon as a function parameter enables cleaner outputs.
This also simplifies our ability to use the data changed flag to send data to microprocessors (without flooding them with unused data).

```rust
conn.add_data_definition(0, "PLANE LATITUDE", "Degrees", simconnect::SIMCONNECT_DATATYPE_SIMCONNECT_DATATYPE_FLOAT64, u32::MAX, 0.0); // this will trigger an update every minor change < 1.0 degree
conn.add_data_definition(0, "PLANE LONGITUDE", "Degrees", simconnect::SIMCONNECT_DATATYPE_SIMCONNECT_DATATYPE_FLOAT64, u32::MAX, 1.0); // this will only trigger an update every 1.0 degree change```

BitsAndDroids committed:
Add epsilon parameter to data definition

An epsilon parameter is added to the data definition in the SimConnect_AddToDataDefinition function in lib.rs. This replaces the previously hard-coded value of 0.0, allowing for more flexibility depending on the specific requirements of different use cases.

BitsAndDroids committed:
upped v

BitsAndDroids committed:
Add optional epsilon parameter to add_data_definition functions

The `add_data_definition` functions in `src/lib.rs` now accept an optional 'epsilon' parameter as opposed to a mandatory float. This modification enhances versatility by accommodating scenarios where 'epsilon' doesn't require a specific value and defaults to 0.0 if not explicitly given.

BitsAndDroids committed:
Updated add_data_definition method to include optional epsilon parameter

The add_data_definition method has been updated to include an optional epsilon parameter. This parameter is used to set a threshold for updates, providing more flexibility for data definition. If no specific value for epsilon is provided, it defaults to 0.0.

BitsAndDroids committed:
Update simconnect dependency in README

The simconnect dependency version in the README file is updated from "0.2.0" to "0.2.1". This reflects the current version the project is using and ensures consistency across all parts of the documentation.

BitsAndDroids committed:
Update unsafe block code in Readme and lib.rs

The handling of raw pointers and data transmutation in the unsafe blocks has been refactored for better safety and code clarity. Instead of using transmute_copy, it now uses std::ptr::addr_of and std::ptr::read_unaligned to obtain data values.

BitsAndDroids committed:
Update simconnect version and add new examples

Updated `simconnect` package from version `0.2.0` to `0.2.1` in `Cargo.lock` file. Introduced two new example scripts `aircraft_updates_on_change` and `aircraft_updates` to demonstrate usage of the libraries. Also, made a minor change in `lib.rs` on how `epsilon` is handled inside `add_data_definition` function.

BitsAndDroids committed:
Improve data update definition in aircraft example

Updated how data is defined in aircraft_updates_on_change example. Epsilon is now defined to control the frequency of updates from the game, reducing the amount of data sent to the client. For instance, latitude and longitude values update every degree, whereas altitude updates are now received every 100 feet, as opposed to every foot previously.

BitsAndDroids commented:
Added an extra example to explain how to utilize the epsilon variable and put both examples in the example folder

BitsAndDroids committed:
Update aircraft data request to prevent overwrite

Added comments in the aircraft_updates_on_change example to explain how different requests need to have unique request_id values in order to prevent data overwrite. Also included a pointer about the way sim_data_ptr uses the passed define_id for better clarity.

BitsAndDroids committed:
Add structured definitions to aircraft example

Expanded on the usage of KeyValuePairFloat and the need for unique request_id values in the aircraft_updates_on_change example. Clarified how define_id is used with simconnect data by adding detailed comments, helping prevent potential overwrite of data.

Sequal32 commented on Readme.md:
```diff
@@ -40,8 +40,10 @@ loop {
-            let sim_data: DataStruct = transmute_copy(&data.dwData);
-            println!("{:?} {:?} {:?}", sim_data.lat, sim_data.lon, sim_data.alt);
+            let sim_data =  std::ptr::addr_of!(data.dwData);
+            let sim_data_ptr = sim_data as *const DataStruct;
```
We can save some lines by making these one line

```suggestion
            let sim_data_ptr =  std::ptr::addr_of!(data.dwData) as *const DataStruct;
```

Sequal32 commented on examples/aircraft_updates_on_change/main.rs:
```diff
@@ -0,0 +1,108 @@
+use std::time::Duration;
+use std::thread::sleep;
+
+// To allign the memory we have to set a fixed max size to the returned variables from the game
```
Typo

```suggestion
// To align the memory we have to set a fixed max size to the returned variables from the game
```

Sequal32 commented on examples/aircraft_updates_on_change/main.rs:
```diff
@@ -0,0 +1,108 @@
+    conn.connect("Program that returns data on changes"); // Intialize connection with SimConnect
+
+    // Here we define all our variable that get returned as floats
+    // (including integers, the memory allignment will handle the)
```
```suggestion
    // (including integers, which the memory alignment will handle)
```

Sequal32 commented on examples/aircraft_updates_on_change/main.rs:
```diff
@@ -0,0 +1,108 @@
+            Ok(simconnect::DispatchResult::SimobjectData(data)) => {
+                unsafe {
+                    match data.dwDefineID {
+                        //Here we match the define_id we've passed using the request_data_on_sim_object
```
```suggestion
                        // Here we match the define_id we've passed using the request_data_on_sim_object
```

Sequal32 commented on examples/aircraft_updates_on_change/main.rs:
```diff
@@ -0,0 +1,108 @@
+                            // The amount of floats received from the sim
+                            let count = data.dwDefineCount as usize;
+
+                            // itterate through the array of data structs
```
```suggestion
                            // Iterate through the array of data structs
```

Sequal32 commented on examples/aircraft_updates_on_change/main.rs:
```diff
@@ -0,0 +1,108 @@
+                            let count = data.dwDefineCount as usize;
+
+                            // itterate through the array of data structs
+                            // To allign the memory we have allocated an array of 255 elements to the datastruct
```
```suggestion
                            // To align the memory we have allocated an array of 255 elements to the datastruct
```

Sequal32 commented on examples/aircraft_updates_on_change/main.rs:
```diff
@@ -0,0 +1,108 @@
+                            // itterate through the array of data structs
+                            // To allign the memory we have allocated an array of 255 elements to the datastruct
+                            // The game might return 255 or 2 values
+                            // To only itterate over valid elements in the array
```
```suggestion
                            // To only iterate over valid elements in the array
```

Sequal32 committed:
Update Readme.md

Sequal32 reviewed (changes requested):
Thanks so much for your contribution and including examples with detailed descriptions! These will make it much easier for developers to start using the crate and know how to read floats/strings. 

Apologies for not including the epsilon value in the function call. I commented on a few typos and a suggestion to combine some lines. I also committed an update to `Readme.md` with some changes inspired by #9

BitsAndDroids committed:
Update code comments and fix typographical issues

This commit rectifies typographical errors and improves the readability of code comments in both the README file and example files. Detailed instructions regarding the execution of examples with SimConnect.dll have been included. Additionally, the datastruct memory allocation has been adjusted for optimal performance and accuracy.

BitsAndDroids committed:
Merge branch 'master' of https://github.com/BitsAndDroids/simconnect-rust

# Conflicts:
#	Readme.md

BitsAndDroids committed:
Update terminology in Readme.md

This commit corrects typographical errors in the Readme.md file. It also enhances the instructions on building the code and running the examples to help users understand them easily. Hyphenations are also updated to ensure grammatical correctness.

BitsAndDroids commented:
@Sequal32 Whoops, I just started using RustRover and haven't configured my spellchecker yet. I've amended the spelling errors (and found some more). 
Over time, I will add some extra examples, such as communication with in-game WASM modules.

PS. Happy holidays

BitsAndDroids committed:
Fix typo in connect function comment in main.rs

This commit corrects a typographical error in the inline comment when connecting with SimConnect in main.rs within the aircraft_updates_on_change directory. Instead of "Intialize", the correct spelling "Initialize" is now used.

Sequal32 committed:
fix: cleanup clippy warnings

Sequal32 commented:
Awesome, the comments look great. I ran `cargo clippy` over the whole project and fixed those warnings (and added print statements for the `Open` and `Quit` events as well). Thanks so much for your contribution and happy holidays!

Sequal32 committed:
refactor: renamed SimobjectData -> SimObjectData

Sequal32 committed:
chore: increment breaking change version num

MrMinimal commented:
I updated to the latest version and saw all these changes. Superb contribution, the examples will be amazing for new devs getting into this. Thank you!
